middleware.js

```diff
@@ -0,0 +1,211 @@
+/**                   _
+ *  _             _ _| |_
+ * | |           | |_   _|
+ * | |___  _   _ | | |_|
+ * | '_  \| | | || | | |
+ * | | | || |_| || | | |
+ * |_| |_|\___,_||_| |_|
+ *
+ * (c) Huli Inc
+ */
+
+
+/**
+ * @file Manages and executes middlewares in all the different app phases
+ * @requires ju-shared/observable-class
+ * @module ju-mvc/middleware
+ * @extends ju-shared/observable-class
+ */
+
+define([
+	'ju-shared/observable-class'
+],
+function (
+	ObservableClass
+){
+	'use strict';
+
+
+	/** Constants **/
+
+    /**
+     * @constant {String} DEFAULT_SUBPHASE - if the subphase is not provided when adding a middleware this will be use
+     */
+	var DEFAULT_SUBPHASE = 'during';
+
+    /**
+     * Object used to store all the phases and middlewares
+     * @type {Object}
+     */
+    var phases = {};
+
+
+    /**
+     * Create an instance of middleware
+     * @constructor
+     * @alias module:ju-mvc/middleware
+     * @returns {Object} Middleware
+     */
+	var Middleware = ObservableClass.extend({
+
+
+		init : function(){
+            //initialize the phases
+            var pashesKeys = Object.keys(Middleware.PHASES);
+            for (var currentPhase = 0, pashesTotal = pashesKeys.length; currentPhase < pashesTotal; currentPhase++) {
```

Comment:
Thinking that maybe phases and subphases should be more related. I wouldn´t be able to define a RENDER phase with unique subphases like PRE_RENDER , VIEW_MODEL, etc

Or maybe this class should be an abstract for middleware, with a child implementation for routing. Phases and subphases can be define at Route.js class members and the middleware can be attached to the routing. Lets discuss it